### ISSUE TITLE
Warning: dereferencing type-punned pointer will break strict-aliasing rules

### ISSUE THREAD

**Issue by lgeretti** [ariadne-cps/ariadne#130]
Compiling with -O3 with gcc issues a bunch of strict-aliasing warnings.

**Comment by pietercollins** [comment]
I looked into this issue in the past, but forget some of the details.
It *should* be safe to cast a block of memory (new[] array of byte/char) to a class type and use placement-new; I don't know if this idiom triggers these warnings.
We used to reinterpret_cast the various MultiIndex and ExpansionValue Reference and Pointer classes, but we should just use static_cast now, and I *think* this should not trigger these warnings, but I'm not sure.
In any case, I think it's best to remove the causes of these warnings if possible, but we should take a look at what's causing them first and see if it's an idiom we want to keep.
